contracts/liquidity-protection/LiquidityProtection.sol

```diff
@@ -639,42 +632,145 @@ contract LiquidityProtection is ILiquidityProtection, Utils, Owned, ReentrancyGu
+
+        // compensate the caller with network tokens if still needed
+        uint256 delta = _networkCompensation(targetAmount, baseBalance, packedRates);
+        if (delta > 0) {
+            // check if there's enough network token balance, otherwise mint more
+            uint256 networkBalance = _networkToken.balanceOf(address(this));
+            if (networkBalance < delta) {
+                _networkTokenGovernance.mint(address(this), delta - networkBalance);
             }
-            _networkV3.migrateLiquidity{ value: value }(
+
+            // lock network tokens for the caller
+            _networkToken.safeTransfer(address(_wallet), delta);
+            _lockTokens(provider, delta);
+        }
+
+        // if the contract still holds network tokens, burn them
+        uint256 networkBalance = _networkToken.balanceOf(address(this));
+        if (networkBalance > 0) {
+            _burnNetworkTokens(removedPos.poolToken, networkBalance);
+        }
+    }
+
+    /**
+     * @dev migrates a set of position lists to v3
+     *
+     * Requirements:
+     *
+     * - the caller must be the owner of all of the positions
+     */
+    function migratePositions(MigrationUnit[] calldata migrationUnits) external nonReentrant {
+        for (uint256 i = 0; i < migrationUnits.length; i++) {
+            migratePositions(migrationUnits[i]);
+        }
+    }
+
+    /**
+     * @dev migrates a list of positions to v3
+     *
+     * Requirements:
+     *
+     * - the caller must be the owner of all of the positions
+     */
+    function migratePositions(MigrationUnit calldata migrationUnit) internal {
+        IDSToken poolToken = migrationUnit.poolToken;
+        IReserveToken reserveToken = migrationUnit.reserveToken;
+
+        uint256 poolAmount = 0;
+        uint256 reserveAmount = 0;
+        uint256 targetAmount = 0;
+
+        for (uint256 i = 0; i < migrationUnit.positionIds.length; i++) {
+            Position memory removedPos = _removePosition(msg.sender, migrationUnit.positionIds[i], PPM_RESOLUTION);
+            require(
+                removedPos.poolToken == poolToken && removedPos.reserveToken == reserveToken,
+                "ERR_INVALID_MIGRATION_UNIT"
+            );
+
+            // get the pool token amount
+            poolAmount = poolAmount.add(removedPos.poolAmount);
+
+            // get the reserve token amount
+            reserveAmount = reserveAmount.add(removedPos.reserveAmount);
+
+            // get the various rates between the reserves upon adding liquidity and now
+            PackedRates memory packedRates = _packRates(
+                removedPos.poolToken,
                 removedPos.reserveToken,
-                provider,
+                removedPos.reserveRateN,
+                removedPos.reserveRateD
+            );
+
+            // verify rate deviation as early as possible in order to reduce gas-cost for failing transactions
+            _verifyRateDeviation(
```

Comment:
We can move this check outside of the loop (i.e., it's enough to check that the pool is unstable once before proceeding with any migrations)

Reply:
The wasteful part is when `_reserveTokenRates` is called from `_packRates` (for each position).
That one could possibly be optimized to execute once for all positions, but we'll probably need to implement separate versions for both `_packRates` and `_reserveTokenRates`.
Function `_verifyRateDeviation` itself will still need to be called for each position, but that's pretty negligible in terms of gas-consumption.

Reply:
A quick check on that:
Seems to work.
Contract size increases to 24595 (19 bytes above the maximum), so we will need to find a workaround.

Reply:
Fixed

Reply:
`_verifyRateDeviation` doesn't need to be called once for each position. I think that you misunderstand its purpose.

re: contract size - you can deduplicate back `removeLiquidity` for both of the cases

Reply:
You're right about `_verifyRateDeviation `.
It's pretty insignificant in terms of gas though, but now that I moved the more significant piece (a part of `_packRates`) outside of the loop, I can also move this one pretty easily.

Reply:
good, I think it's better to only call it once in general